schema-layer/data-structures/hashmap.md

```diff
@@ -0,0 +1,299 @@
+
+By default, each node in an IPLD HashMap is stored in a distinct IPLD block and CIDs are used for child node links. The schema and algorithm presented here also allows for inline child nodes rather than links, with read operations able to traverse multiple nodes within a single block where they are inlined. The production of inlined IPLD HashMaps is left unspecified and users should be aware that inlining breaks canonical form guarantees.
+
+## Structure
+
+### Parameters
+
+Configurable parameters for any given IPLD HashMap:
+
+* `hashAlg`: The hash algorithm applied to keys in order to evenly distribute entries throughout the data structure. The algorithm is chosen based on speed, `digestLength` and randomness properties (but it must be available to the reader, hence the need for shared defaults, see below).
+* `bitWidth`: The number of bits to use at each level of the data structure for determining the index of the entry or a link to the next level of the data structure to continue searching. The equation `2`<sup>`bitWidth`</sup> yields the arity of the HashMap nodes, i.e. the number of storage locations for buckets and/or links to child nodes.
+* `bucketSize`: The maximum array size of entry storage buckets such that exceeding `bucketSize` causes the creation of a new child node to replace entry storage.
+
+### Node properties
+
+Each node in a HashMap data structure contains:
+
+* `data`: An Array, with a length of one to `2`<sup>`bitWidth`</sup>.
+* `map`: A bitfield, stored as Bytes, where the first `2`<sup>`bitWidth`</sup> bits are used to indicate whether a bucket or child node link is present at each possible index of the node.
+
+An important property of a HAMT is that the `data` array only contains active elements. Indexes in a node that do not contain any values (in buckets or links to child nodes) are not stored and the `map` bitfield is used to determine the `data` whether values are present and the array index of present values using a [`popcount()`](https://en.wikipedia.org/wiki/Hamming_weight). This allows us to store a maximally compacted `data` array for each node.
+
+### Schema
+
+The **root block** of an IPLD HashMap contains the same properties as all other blocks, in addition to configuration data that dictates how the algorithm below traverses and mutates the data structure.
+
+See [IPLD Schemas](/schema-layer/schemas) for a definition of this format.
+
+```ipldsch
+# Root node layout
+type HashMapRoot struct {
+  hashAlg String
+  bitWidth Int
+  bucketSize Int
+  map Bytes
+  data [ Element ]
+}
+
+# Non-root node layout
+type HashMapNode struct {
+  map Bytes
+  data [ Element ]
+}
+
+type Element union {
+  | Child "child"
+  | Bucket "bucket"
+} representation keyed
+
+type Child union {
+  | Link link
+  | HashMapNode map
+} representation kinded
+
+type Bucket list [ BucketEntry ]
+
+type BucketEntry struct {
+  key Bytes
+  value Value (implicit "null")
+} representation tuple
+
+type Value union {
```

Comment:
This whole thing might be possible to replace with the `Any` type.

(Which... needs specification.  But it's come up more than a few times that we need a way to go back to free-for-all untyped data land, and when that's been happening, I've been calling it `Any`.  So I'd say feel free to do so here, too -- and we should go sort out a full spec for `Any` soon.)

Reply:
👍 my original draft used `any` as a placeholder but since we don't have that defined yet ...